pkg/s3select/select.go

```diff
@@ -338,24 +356,29 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
 	var inputRecord sql.Record
 	var outputRecord sql.Record
 	var err error
-	var data []byte
+	bufW := bufio.NewWriter(ioutil.Discard)
```

Comment:
I don't understand the need of `bufW` i.e. `bufio.Writer` to `ioutil.Discard`. As you are using bytes.Buffer` which is already an buffer, why can't we just use it?

Reply:
this is a little tricky,  here we use bufio Writer to prevent csv.Writer from allocating new buffer.

in csv.NewWriter, if w is not  the type of bufio.Writer, defaultBufSize slice will be maked:
https://github.com/golang/go/blob/master/src/encoding/csv/writer.go#L40
https://github.com/golang/go/blob/master/src/bufio/bufio.go#L556

Reply:
Looks like a bug in csv writer in go standard library. It could use as is if given writer is `bytes.Buffer` like `bufio.Writer`.  Having `bufio.Writer` does not give any value here.  You could use `bytes.Buffer` as is and let csv writer in go standard library wrap with `bufio.Writer`.

Reply:
If we let csv writer in go standard library wrap with bufio.Writer, https://github.com/yaozongyou/minio/blob/speed-up-select-on-csv/pkg/s3select/csv/record.go#L65 will be called for each output record. and defaultBufSize will be maked for each output record,  this will be the performance killer, especially in a concurrent situation.

![image](https://user-images.githubusercontent.com/590761/62346286-25dbec80-b528-11e9-8d44-cc92ed81898f.png)

the above image is the profile.

Reply:
I see you want one `bufio.Writer` per `Evaluate()`.  Due to lake of comments explaining why its done, its hard to understand.

1. You would need to comment `bufioWriter := bufio.Writer(ioutil.Discard)` that its merely for creating `bufio.Writer` and later using `bufioWriter.Reset(w)` is used to set correct destination writer.

2. Now we need two different arguments for output format type. CSV needs `bytes.Buffer` and `bufio.Writer`, whereas JSON needs `bytes.Buffer`. This is little weird.  This can be avoided by having global `bufioWriterPool` like `bufPool` (may be performance improvement for concurrent select requests with CSV output format).

My suggestion is like below
```go
var bufioWriterPool = sync.Pool {
	New: func() interface{} {
		// ioutil.Discard is just used to create the writer. Actual destination writer is set later by Reset() before using it.
		return bufio.NewWriter(ioutil.Discard)
	},
}

func (s3Select *S3Select) marshal(buf *bytes.Buffer, record sql.Record) error {
 	switch s3Select.Output.format {
 	case csvFormat:
 		bufioWriter := bufioWriterPool.Get().(*bufio.Writer)
 		defer func() {
			bufioWriter.Reset(ioutil.Discard)
			bufioWriterPool.Put(bufioWriter)
		}()

 		bufioWriter.Reset(buf)
 		err := record.WriteCSV(bufioWriter, []rune(s3Select.Output.CSVArgs.FieldDelimiter)[0])
		if err != nil {
			return err
		}

		buf.Truncate(buf.Len() - 1)
		buf.WriteString(s3Select.Output.CSVArgs.RecordDelimiter)
		return nil

	case jsonFormat:
		...
		...
		...
	}
	...
}
```

Reply:
good idea. 
but by the way, we can not use `bufio.NewWriterSize(ioutil.Discard, 1)` to avoid creating underneath buffer, because csv.NewWriter wrap our writer with defaultBufSize, and  bufio internally will check and make sure the size is at lease defaultBufSize.
https://github.com/golang/go/blob/master/src/encoding/csv/writer.go#L40
https://github.com/golang/go/blob/master/src/bufio/bufio.go#L554

Reply:
changed